proposal/proposal.tex

```diff
@@ -225,50 +211,77 @@ \subsection{Hygiene}
 Point \ref{responsibility}. above touches on an important topic: PPXs generate
 the final representation of the program and so have an impact on how the
 compiler and the editor tool $merlin$ analyze the program. That's why we define
-clear hygienic rules\cite{hygiene} and try to create a culture in which PPX
-authors are aware of the possible consequences.
+clear hygiene rules\cite{hygiene} and try to create a culture in which PPX
+authors are aware of the possible consequences. Some of these hygiene rules are
+enforced technically by $ppxlib$, and others are currently only guidelines that
+PPX authors should follow. The PPX mechanism within the compiler itself does not
+enforce any technical notion of hygiene as in some macro
+systems\cite{scheme_hygiene}.
 
 \subsubsection*{Error handling}
 
-It's important to handle PPX errors at the level of the AST abstraction.
-Concretely, errors need to be embedded as nodes into the AST. A raising PPX
-means that the compiler and $merlin$ don't receive an expanded AST.
-%and won't be able to reason about the whole program.
+Errors in OCaml are most often handled by raising exceptions. Naive error
+handling within the $ppxlib$ driver would abort transformation of an entire
+program the first time an exception was raised. This means tools like $merlin$
+learn nothing about the shape of the transformed program, and debugging the
+problem becomes only more difficult.
+
+Our first approach to addressing this was to provide tools for expressing errors
+as AST nodes, and encouraging PPX authors to use these instead of exceptions.
+This practice was slow to creep into the ecosystem, so we made the behavior
+automatic by handling exceptions inside $ppxlib$. Now an exception raised by a
+context-free transformation is caught, and its error message is stored as an AST
+node so that the rest of the program can be transformed.
```

Comment:
Arrg, thanks for noticing and correcting this! We should make clear that raising interrupts the rest of the rewriting, though, and hence it's still recommended to embed errors instead of raising. Maybe something along the lines of
```suggestion
Errors in OCaml are most often handled by raising exceptions. Naive error
handling within the $ppxlib$ driver would abort transformation of an entire
program the first time an exception was raised. This means tools like $merlin$
learn nothing about the shape of the transformed program, and debugging the
problem becomes only more difficult.

Our first approach to addressing this was to provide tools for expressing errors
as AST nodes, and encouraging PPX authors to use these instead of exceptions.
This practice was slow to creep into the ecosystem, so we made the behavior
automatic by handling exceptions inside $ppxlib$. Now an exception raised by a
context-free transformation is caught, and its error message is stored as an AST
node inside the last AST of the rewriting process. So a raising PPX still means that
the rewriting flow is interrupted, and the `ppxlib` driver returns a partly rewritten
AST. Therefore, we still encourage PPX authors to embed errors instead of raising.
```
? What do you think?

Reply:
What stage of the rewriting is interrupted? The closest enclosing context-free rewrite? Or the whole thing?

Reply:
The stage of the latest full AST. So in the case of a raising context-free rewriter, the AST from before the context-free phase.

Reply:
Oh. Hmm. We may want to consider catching things at the level of individual extension nodes.